test/viper/ok/record.vpr.ok

```diff
@@ -0,0 +1,125 @@
+/* BEGIN PRELUDE */
+/* Array encoding */
+domain Array {
+  function $loc(a: Array, i: Int): Ref
+  function $size(a: Array): Int
+  function $loc_inv1(r: Ref): Array
+  function $loc_inv2(r: Ref): Int
+  axiom $all_diff_array { forall a: Array, i: Int :: {$loc(a, i)} $loc_inv1($loc(a, i)) == a && $loc_inv2($loc(a, i)) == i }
+  axiom $size_nonneg { forall a: Array :: $size(a) >= 0 }
+}
+define $array_acc(a, t, p) forall j: Int :: 0 <= j && j < $size(a) ==> acc($loc(a, j).t, p)
+define $array_untouched(a, t) forall j: Int :: 0 <= j && j < $size(a) ==> $loc(a, j).t == old($loc(a, j).t)
+define $array_init(a, t, x) forall i : Int :: {$loc(a, i).t} 0 <= i && i < $size(a) ==> $loc(a, i).t == x
+/* Tuple encoding */
+adt Tuple$2 [T0, T1] { Tup$2(tup$2$0 : T0, tup$2$1 : T1) }
+/* Option encoding */
+adt Option[T] {
+  None()
+  Some(some$0: T)
+}
+/* Text encoding */
+function $concat(a: Int, b: Int): Int
+/* Typed references */
+field $c_R: R
+/* END PRELUDE */
+
+define $Perm($Self) ((((true && acc(($Self).fld1,write)) && acc(($Self).fld2,write)) && 
+  acc(($Self).empty,write)))
+define $Inv($Self) (true)
+method __init__($Self: Ref)
+    
+    requires $Perm($Self)
+    ensures $Perm($Self)
+    ensures $Inv($Self)
+    { 
+      ($Self).fld1 := $RecordCtor_R(42, 3);
+      ($Self).fld2 := $RecordCtor_R(42, 3);
+      ($Self).empty := $RecordCtor_R(0, 4); 
+    }
+adt R  { $RecordCtor_R(a : Int, b : Int) }
+field fld1: R
```

Comment:
I see here that if an actor have field `a` then Viper doesn't parse translation outcome. Or if there are two types `type A = {x: Int}` and `type B = {x: Bool}` that probably fine in Motoko.

Thus I think it make sense to have something like:
```suggestion
adt R  { $RecordCtor_R($R$a : Int, $R$b : Int) }
field fld1: R
```
And deduce name of the Viper field from the expression type. So we can keep some map in the `ctxt` or just use a pure function `record type name, field name -> Viper field name`

Reply:
BTW, could `type_to_record_ctor` be just a pure function? 
For example for variants we did it like that without state

Reply:
> BTW, could type_to_record_ctor be just a pure function?
For example for variants we did it like that without state

Motoko doesn't store original type name in its types. It works for variants because we can easily deduce the Viper's ctor name from the Motoko's one. Records don't have an explicit one, so, we should store this map in the context.